### ISSUE TITLE
In game guide book

### ISSUE THREAD

**Issue by Lilylicious** [sinkillerj/ProjectE#823]
Opening the issue to keep track of guidelines/requests/thoughts.

1. No mod dependency.
2. Index support.
3. Basic image support.
4. Attempt to keep page adding/sorting/management as easy and streamlined as possible.

**Comment by williewillus** [comment]
Perhaps a way to read from an online resource such as a git repo (so we can make live edits)

Maybe some way to incorporate the current wiki (which i should get back to work on)

**Comment by sinkillerj** [comment]
I dont see the need for online reading simply because any text is going to be specific to the version of the mod and what it includes.

**Comment by Lilylicious** [comment]
Online reading adds another layer of complexity that is frankly unnecessary. The main question for me is the approach: Menu or text heavy.

Botania is an example of a menu-heavy book, where you click a category that contains pages that you click to open. It makes the guide less limited into a traditional book structure and unlocks more freedom in what type of pages you add, whether it be text or big pictures or recipe pages and whatnot.

RFTools on the other hand follows a much more text-heavy structure and is more akin to a real book, where page 2 follows page 1 etc. With a table of contents with links and a searchable index, this can be as easy to navigate as the menu based system while allowing for a more natural book feel.

**Comment by sinkillerj** [comment]
I would like to see the index be like:
**Category Name**
Link
Link
Link

**Comment by sinkillerj** [comment]
This may result in the index being multiple pages, but I dont feel that each category needs to link to its own page of links.

**Comment by Lilylicious** [comment]
Do we want to keep pages linked together traditional book style? Turning pages in other words. Could be useful for misclicks, simply turn the page to the right one.

**Comment by sinkillerj** [comment]
Yea the index just linking to a page number would be nice so you can flip through, but you could always click "return to index" at any time.

**Comment by Lilylicious** [comment]
@williewillus, regarding the location for the files: If we want to keep it gathered, it can all go in gameObjs.items.manual or something to that nature. Another option is to divide up the code for the manual (previously mentioned package) and put pages somewhere else, perhaps even in assets depending on how I implement things.

**Comment by williewillus** [comment]
Just do whatever for now, I find it easier to reorganize as it develops than trying to factor it all in at the beginning before anythings written :)

**Comment by Lilylicious** [comment]
After doing research and thinking, I've decided on how to proceed (for now) in regards to the page infrastructure.

There will be a directory of txt files from which the mod will grab and fill arrays. These will then assign their index to a series of constant ints to make the code more readable and fluid.

The details will come, that's the gist of it. Don't have an ETA, but I'm working on it.

**Comment by williewillus** [comment]
Make sure to keep localizability in mind

**Comment by Lilylicious** [comment]
Noted.

**Comment by williewillus** [comment]
Yeah, I don't think you can look up lang keys out side of the lang files, so Botania just sticks all of the lexicon text in there (https://github.com/Vazkii/Botania/blob/master/src/main/resources/assets/botania/lang/en_US.lang)

**Comment by Lilylicious** [comment]
Dislike the system, but if no other options exist for localization we'll have to make do.

**Comment by Lilylicious** [comment]
The text will be in the lang files, found a way to make it neat still.

**Comment by Lilylicious** [comment]
@williewillus @sinkillerj I'm making great progress, wanted some input before I went ahead with this:

Each page needs to be registered somehow. One good option is to do it at the same time as registering the item/block itself to the gameregistry through a function that does both.

Doing it that way would ensure that future blocks added to the gameregistry through the function will automatically get a page added to the guide book.

Another way that has a lower impact on the current code is to do a class for registering all the blocks we want in the guide as pages. Thoughts?

**Comment by williewillus** [comment]
I'd prefer the second one - just one kinda "PageHandler" class that takes stuff to register.
I want to minimize how closely tied this will be to the existing codebase (modularization -> good)

**Comment by SnapshotDesign** [comment]
when do you believe that the new update with the bug fixes will be releases?

**Comment by williewillus** [comment]
@SnapshotDesign please keep your comments on topic to this ticket.
And we just released a bugfix update, 1.7.1. If you have anymore problems please report them separately

**Comment by williewillus** [comment]
Also, is this just the framework or will it have the entries too? I want to loreify all the entries (basically move the wiki into the book lol)

**Comment by Lilylicious** [comment]
The text content will be in lang files, the item/block icon is auto-added as is the name. Further picture support is not yet present.

**Comment by williewillus** [comment]
kk!
I have quite a few of the wiki pages written already so after you're done I'll go ahead and import content and work on writing the rest (since I'm apparently Lore Lord or something like according to sin lol)

**Comment by Lilylicious** [comment]
After I get everything working I'll push to my forked branch so you can start working on the lang file entries. The table of contents still needs a while before it'll be ready, so won't pull for another day or two.

**Comment by Lilylicious** [comment]
First pushed commit is now in my branch. Link to the commit: https://github.com/Lilylicious/ProjectE/commit/69a531baf191b563cd83917817582ce87a755399

It's lacking a ToC and searchable index, but basic functionality is there. I'm sure there's cleanup I've missed too, but it's in a pretty stable state.

The gui is something I mocked up, basically a placeholder. The guide book doesn't have an icon yet.

The recipe is something I put in just to get one in, it's not final (shapeless book+dark matter). @sinkillerj , thoughts on that recipe/alternates?

As it stands, using the book doesn't pause the game in singleplayer, though it can (config?).

Adding pages to the manual is done in two steps:
1) Add the item/block to the addPages funciton in ManualPageHandler.java. I've used the variables in ObjHandler to do it.
2) Add text in a lang file where the key is help.unlocalizedName.info, example is present at the bottom of en_US.lang

Adding the item to ManualPageHandler.java makes a page appear with the items/blocks icon and localized name. If the lang file lacks the text, it will simply print the help.unlocalizedName.info key.

To sum up:
1) ToC/Index is missing in this version
2) GUI is a placeholder, guide book icon is missing
3) Recipe is a placeholder, book+dark matter
4) Maybe add a config for game pausing when reading book in singleplayer?
5) Thoughts on the page adding process?

**Comment by williewillus** [comment]
eh I feel like DM is really really expensive for what's supposed to be an entry-level guide, idk.

**Comment by Lilylicious** [comment]
Agreed, but I just threw it in there for placeholder sake. It's meant to change with input.

**Comment by sinkillerj** [comment]
I will be changing the recipe, atm just make a placeholder. I will also be giving new players to the server a copy.

**Comment by Lilylicious** [comment]
Yup, placeholder for now is Book + Dark Matter.

**Comment by Lilylicious** [comment]
To document things said in IRC: Some minor issues arose in regards to items in the PR from willie, the feature is not yet code-complete and in the near future will be added upon to make it do everything we want to (index, non-item bound pages, images etc)

**Comment by Lilylicious** [comment]
Non-item pages are done. Index next, then picture support.

Question: Pictures can either be on the same page or on the next page in line. What're your thoughts on that? Having a separate page would allow bigger pictures. Maybe have both as options?

**Comment by williewillus** [comment]
Given the small size of the spread, I vote next page. Same page is really pressuring on the amount of text you can put, and I write quite a bit :p
Though some entries are quite brief, so maybe having both options is wise. 
Add support for a little caption under the pic

**Comment by Lilylicious** [comment]
I'll do next page as the default option, then see if I can't put in support for an optional same page version.

Caption request noted.

**Comment by MaPePeR** [comment]
Can we make it possible that Items are displayed in a way the NEI R and U key work?

**Comment by Lilylicious** [comment]
No promises, but I'll look into it. Would require actually putting the itemstack there rather than just rendering the icon, I think.

**Comment by Lilylicious** [comment]
Basic image support: https://github.com/Lilylicious/ProjectE/commit/84e40da353bbd22f2623a69d6e1d4f9da50810f1

Allows for imagepages with a title and image. Images need to be 256x256.

Support for in-page images for other pagetypes will be looked at.

**Comment by Lilylicious** [comment]
@sinkillerj @williewillus @MaPePeR  I'm considering generating files manually. https://github.com/Lilylicious/ProjectE/blob/manual/src/main/java/moze_intel/projecte/manual/ManualPageHandler.java#L42 is the current implementation for registering pages, a for loop containing all pages of that type being generated and added to the list of pages.

Instead, I'm considering doing each manually. Something like
```
pages.add(new PEManualPage(ObjHandler.blackHole));
pages.add(new PEManualPage(ObjHandler.voidRing));
```
This would allow for much higher fine control regarding page ordering, as well as allowing me to easily register the page number for index purposes.

I'm going to go ahead and start on this implementation, if anyone has any protests shout at me.

**Comment by sinkillerj** [comment]
+1 for manual

**Comment by Lilylicious** [comment]
Items/Blocks are now added by using
```
addItems(ObjHandler.repairTalisman);
```
It handles both blocks and items, takes in Item/Block, passes that to PEManualPage constructor that makes it into itemstacks and everyone's happy.

**Comment by Lilylicious** [comment]
The index is harder to do than I anticipated. Researching options now.

I have the pages done, and the ability to order them as you want, as well as a HashMap of the unlocalized name (pe_mercurial_eye for example) as a key with the page index as the value.

The problem is how to add them to the index page. I could do "button" invisible rectangles over text, could potentially work. We'll see, research time.

**Comment by williewillus** [comment]
Cheat off botania :p

**Comment by Lilylicious** [comment]
It's a pain in the bum to sort through, but it's a good idea.

**Comment by Lilylicious** [comment]
Okay. Okay. I did it. Sort of. The index works. Kind of.

The problem lies in the openGL scaling. As it stands, the size works visually, but the links are offset. I'm pushing the code to my repo, any thoughts and suggestions are welcome. I'll go to sleep now, get back on it tomorrow with fresh eyes and fix it.

**Comment by Lilylicious** [comment]
Never, ever, code, when tired. https://github.com/Lilylicious/ProjectE/commit/8172e12f28159699a9223854553cf1fde0dc8536

The positioning looks a bit weird, but I'll take it. Fix tomorrow.

Before that commit: http://i.imgur.com/Zrsv0n9.png After: http://i.imgur.com/7MsgwlP.png

**Comment by Lilylicious** [comment]
Due to my late night coding shenanigans, I'm in no shape to write halfway decent code today, so taking the day off.

I want to add in categories to the index screen, or at least a way to put them in. Also want to take a look at the scaling, I might be able to multiply some things to make the pixels actually make sense.

After that I'll need more input from y'all in regards to what categories are wanted and/or content for the pages. Do you want to take care of that before the book is merged into the main project? You can do pull requests to my manual branch for that, if you wish, then pull into the main project when we're happy with the end result.

**Comment by williewillus** [comment]
I have most of the articles written (from the wiki), I'll throw up a google docs here in a bit so we can do the rest as we go

**Comment by williewillus** [comment]
https://docs.google.com/document/d/1N0R1dbiiik50C_sYFlX64aZPtAdI4NT4Cxmt-rp-Fxg/edit?pli=1

**Comment by sinkillerj** [comment]
@Lilylicious I can categorize after its merged.

**Comment by Lilylicious** [comment]
@sinkillerj @williewillus The latest commit is up on my branch and it's a doozy, 282 additions and 221 deletions. (Refactored a bit, that's the reason for the deletions).

I expect the index to "break" when we have all the categories in, so there's surely more work to be done there. Adding categories is as simple as adding a case to a switch statement with it. Now that I look at it, I forgot to do localization there, but that's a quick fix.

Some example pictures here: http://imgur.com/a/f87ph The Death category looks a bit out of place, which is fixed by altering the category structure or by adjusting the yOffset cutoff for new line (at https://github.com/Lilylicious/ProjectE/blob/98a8f5f0b15158ebda49a007fb0ddab8445f1227/src/main/java/moze_intel/projecte/gameObjs/gui/GUIManual.java#L275)

I'm waiting on @williewillus to give it a lookover for feedback/refactoring, but we're rapidly approaching a pull request.

@sinkillerj, do you want to finalize it before we merge, or merge when we're just missing the content? Could do work on my branch and pull request into that until it's ready, then move to the main project. We could use willie's google doc for deciding and cementing the categories/page order, which I can then do in the code. Or willie/you PR that. Or we PR then merge then fix in main project.

**Comment by MaPePeR** [comment]
Whatever you do. Don't try to fix this code on `master`

**Comment by Lilylicious** [comment]
Of course not.

**Comment by sinkillerj** [comment]
@Lilylicious Remind me next time I am on IRC and I will give it a look over.

**Comment by Time6628** [comment]
Idk if this helps at all but this is a pretty cool mod/API that allows you to create in-game documentation easier [[Link]](http://www.curse.com/mc-mods/minecraft/228832-guide-api) :stuck_out_tongue:

**Comment by sinkillerj** [comment]
@Time6628 A api has been discussed, and will not be used at this time. A api would require a additional mod download which would instantly prevent those who didn't know from using the guide, thus defeating the purpose.

**Comment by sinkillerj** [comment]
Major issues I've noticed testing this:
The GUI could take up more space, and the text renders way too small on the default window size: http://i.imgur.com/AR9dJFe.png

In order to avoid this you need to consider that info, including the index, could need to take up multiple pages. The GUI is also still one big awkward page instead of being split into two pages at once like a book.

**Comment by Lilylicious** [comment]
Figured I'd let you guys know, I had a bit of a sensory integration attack-esque event happen (long story), so I was out of commission for the past two or three days. I'm getting back on my feet again, and will get back to finishing this up shortly, but it'll be another few days I think. I know, I know, no rush, just wanted to let y'all know.

Can ask in IRC if yer curious about what happened, if not, don't worry about it. I'm fine and will be finer.

**Comment by Lilylicious** [comment]
Work is resuming on this.

Issues and thoughts:

Bigger GUI - Should be easy enough.
Split pages: Thoughts on this, @sinkillerj? Shouldn't be too hard to accomplish.

Multiple pages per "page" or topic/index. Could use the default text size and  simply add another page with the string split onto the new page where needed, that way I avoid the entire clusterfuck that is glScaling.

**Comment by sinkillerj** [comment]
Multiple pages is def better than scaling, and lets willie let his creative energies out.

**Comment by Lilylicious** [comment]
Multiple pages SHOULD be functional on my fork now. Keyword should.

The problematic line is here: https://github.com/Lilylicious/ProjectE/blob/887a931e5fd13bcd1919c6874ca4507481bb1146/src/main/java/moze_intel/projecte/manual/AbstractPage.java#L16

It's returning as null, where it really really shouldn't. I've tried more things than I can count to try to make it not return null, with no luck.

The reason it exists is to be able to calculate how many pages is necessary while creating the pages, for which it needs to know how many lines a string would be chopped up into. As a result, we need the fontrenderer to make us that list of strings. Problem is, no matter what I do, the fontRenderer is null.

**Comment by Lilylicious** [comment]
Two words: preinit bad.

The multi-page functionality has been mostly finished at this point. The current state of the manual is in the imgur album below.

The index still needs work to be the same as the subpages for regular pages. 

http://imgur.com/a/jOqaw

@sinkillerj @williewillus @MaPePeR Thoughts/Ideas?

**Comment by Lilylicious** [comment]
Work planned for the future: Fix index. Make the actual thing larger, take up more screenspace (mostly wider, a bit taller.)

**Comment by sinkillerj** [comment]
Looking pretty good, still not sold on it being one page shown at a time instead of two like a book though, but if it cant be done it cant be done.

**Comment by Lilylicious** [comment]
It can be done, just hasn't. Another one of those features to maybe be.
Could combine that with the larger gui.

**Comment by sinkillerj** [comment]
Roger, use the transmutation table as a guide height wise, it is the tallest height you can logically go.

**Comment by Lilylicious** [comment]
I expect to have the gui extended by tomorrow. May need to spend some time
tweaking the values, unsure how consistent I was with relative values.

**Comment by Lilylicious** [comment]
First iteration of extended gui/multipage view. http://imgur.com/iUMQmQZ

willie keeps going on about unicode font rendering and whatnot, apparently would be less laggy/take up less space. I'm not sure, but maybe. Lag isn't a problem except on the index page, which I've yet to work on.

**Comment by sinkillerj** [comment]
Looking beautiful, keep up the great work, I have no further complaints atm.

**Comment by williewillus** [comment]
scratch what I said about Unicode font rendering being less laggy, turns out it's the same lagginess as the normal one (bad), but it's smaller.

**Comment by Lilylicious** [comment]
It's very possible that the new index page will have way less of a lag problem, since we'll be drawing a lot less buttons per page. If not, reducing the lag impact of  the index is something we can work on after it's actually working.

**Comment by Lilylicious** [comment]
Lag issue is completely gone. It wasn't due to the fontrenderer as we believed, it was due to a pretty bad coding error on my part. I was creating the index buttons on each frame, instead of in the initgui. That meant that each frame we were displaying 110*framesPassed buttons while on the index. That adds up.

On a positive note, the index is now fully functioning. http://imgur.com/a/1tgHg

What are your thoughts on this? As it stands I'm not using the categories that willie implemented.

@williewillus I haven't used your categories in the current version of the index, it added complexity when I was struggling to make heads or tails of the index as it was. Would you mind working that into your refactoring when you get there?

@sinkillerj Any comments/thoughts on the status of the index?

**Comment by Lilylicious** [comment]
https://github.com/Lilylicious/ProjectE/tree/manual is the current branch. All pushed.

**Comment by williewillus** [comment]
Nice, good catch on the lag.
Yeah that's okay, I'll reevaluate the uses of the categories later, once you like how things look and work outside, poke me and I'll make it pretty inside as well :D

**Comment by Lilylicious** [comment]
I want to extend it downwards before I'm done with it. Other than that I'm pretty happy, want some input from sin before I call it ready.

We also need an icon for the book. I could get us a placeholder, but we could also have someone properly make one. It's not like it'll be released for quite a while, we(you) still have all that content to write.

**Comment by sinkillerj** [comment]
Just put in a placeholder item for now till i can get ahold of it after merge, I will also be adjusting final recipes and adding code for on join granting.

**Comment by Lilylicious** [comment]
I've added a placeholder texture (vanilla book tinted red). I'll extending the gui downwards a bit too soonish.

**Comment by Lilylicious** [comment]
Gui is extended, and fixed a bug where opening the book multiple times in a session would fuck up the index. Not sure what to do next. Any suggestions? You could clone it and play around with the book in game, see if anything pops out as needing improvement.

Considering adding the option to have an image in-page. Not sure how hard it'd be, how it'd interact with text in the page, etc. Might be better/easier to simply use image pages that have a full page image. However, might want to add the option to have an image across both open pages, so basically two image pages that are on the same spread with a single image drawn in the center.

**Comment by Lilylicious** [comment]
MASTER COMMENT FOR KERNING ISSUES:

It has come to my attention that the kerning issues that I noticed through the text is not due to errors in the text itself, but due to the rendering. I'm gathering a list of the problematic parts to see if I can't figure out why it's doing this.

"and refined" (d+f)
"multiple iterations" (e+i)
"a new" (a+n)
"of the" (f+t) (has another spot without the issue)
"grows more" (s+m)

Due to some instances of the words not having this issue, I believe blank spaces are being omitted for some reason.

**Comment by Lilylicious** [comment]
Confirmed. Did a System.out.println for each line with breakpoints and all, the result:

```
[16:17:50] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.gameObjs.gui.GUIManual:drawPage:367]: PEDEBUG:multipleiterations.
[16:17:50] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.gameObjs.gui.GUIManual:drawPage:367]: PEDEBUG:Afterx3n0ph0b3's
[16:17:50] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.gameObjs.gui.GUIManual:drawPage:367]: PEDEBUG:retirement, anew apprentice,
[16:17:50] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.gameObjs.gui.GUIManual:drawPage:367]: PEDEBUG:Pahimar,took up the
[16:17:50] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.gameObjs.gui.GUIManual:drawPage:367]: PEDEBUG:development ofthe art. Our
[16:17:50] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.gameObjs.gui.GUIManual:drawPage:367]: PEDEBUG:story growsmore complex
[16:17:50] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.gameObjs.gui.GUIManual:drawPage:367]: PEDEBUG:after this. Thepeople of the
[16:17:50] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.gameObjs.gui.GUIManual:drawPage:367]: PEDEBUG:countrybegan speaking out,
[16:17:50] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.gameObjs.gui.GUIManual:drawPage:367]: PEDEBUG:shamingalchemists for their
[16:17:50] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.gameObjs.gui.GUIManual:drawPage:367]: PEDEBUG:work,calling them "lazy
[16:17:50] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.gameObjs.gui.GUIManual:drawPage:367]: PEDEBUG:cheaters",uttering strange
[16:17:50] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.gameObjs.gui.GUIManual:drawPage:367]: PEDEBUG:words tothe kin of "EE2
[16:17:50] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.gameObjs.gui.GUIManual:drawPage:367]: PEDEBUG:OP!!!!!!". Theart of alchemy
[16:17:50] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.gameObjs.gui.GUIManual:drawPage:367]: PEDEBUG:itself wasdeemed too
```

**Comment by Lilylicious** [comment]
The StatCollector has the spaces as it should.

So before the drawString but after getBodyText. That leaves the following two spots for the spaces to disappear.

```
  public List<String> getBodyList() {
    return fontRendererObj.listFormattedStringToWidth(getBodyText(), GUIManual.textWidth);
  }
```
```
      bodyTexts = page.getBodyList();

      for (int i = 0; i < bodyTexts.size() && i < Math.floor(GUIManual.textHeight / GUIManual.textYOffset); i++)
      {
        this.fontRendererObj.drawString(bodyTexts.get(i), Math.round(contentX * guiScaleFactor), 43 + textYOffset * i, Color.black.getRGB());
      }

```
bodyTexts.get(i) is missing some spaces, where getBodyText() in the first block is not missing them.

**Comment by MaPePeR** [comment]
its `listFormattedStringToWidth`, then, i guess...

**Comment by Lilylicious** [comment]
Probably, but why? It's not at the start or end of the lines.

**Comment by MaPePeR** [comment]
It might have to do with the formatting that the method tries to parse... i don't know

**Comment by Lilylicious** [comment]
Maybe try running it through a... deformatter..?

**Comment by Lilylicious** [comment]
Noticed a pattern:

On pages where a space goes missing, it happens on EVERY line, except the first.

First page, no issues. Second page, line 1 is fine, line 2+ are missing one space each. Page 3, line one fine, very line after missing a space. Fourth page, same thing.

It's possible that it's omitting a trailing space on the first line, thus having the same problem there but invisibly.

**Comment by Lilylicious** [comment]
Another experiment:

```

  public List<String> getBodyList() {
    System.out.println("PEDEBUG:" + getBodyText());
    List<String> list = fontRendererObj.listFormattedStringToWidth(getBodyText(), GUIManual.textWidth);
    for(String string : list){System.out.println("PEDEBUG:" + string);}
    return fontRendererObj.listFormattedStringToWidth(getBodyText(), GUIManual.textWidth);
  }
```

During startup, getBodyList() is called in the following function in ManualPageHandler.

```
  private static void checkSubPages(AbstractPage page, PageCategory category, String identifier)
  {
    int neededPages = (int) Math.ceil((page.getBodyList().size() * GUIManual.textYOffset) / GUIManual.textHeight);
    int k = 0;
    if(neededPages > 0)
    {
      List<List<String>> parts = chopped(page.getBodyList(), (int) Math.floor(GUIManual.textHeight / GUIManual.textYOffset));
      for (int i = 1; i <= neededPages; i++)
      {
        addSubPage(parts.get(i), category, identifier, k++);
      }
    }
  }
```

The results:
```
[23:55:27] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:33]: PEDEBUG:Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab Ab 
[23:55:27] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:Ab Ab Ab Ab Ab Ab Ab Ab Ab
[23:55:27] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:Ab Ab Ab Ab Ab Ab Ab Ab Ab
[23:55:27] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:Ab Ab Ab Ab Ab Ab Ab Ab Ab
[23:55:27] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:Ab Ab Ab Ab Ab Ab Ab Ab Ab
[23:55:27] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:Ab Ab Ab Ab Ab Ab Ab Ab Ab
[23:55:27] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:Ab Ab Ab Ab Ab Ab Ab Ab Ab
[23:55:27] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:Ab Ab Ab Ab Ab Ab Ab Ab Ab
[23:55:27] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:Ab Ab Ab Ab Ab Ab Ab Ab Ab
[23:55:27] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:Ab Ab Ab Ab Ab Ab Ab Ab Ab
[23:55:27] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:Ab Ab Ab Ab Ab Ab Ab Ab Ab
[23:55:27] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:Ab Ab Ab Ab Ab Ab Ab Ab Ab
[23:55:27] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:Ab Ab Ab Ab Ab Ab Ab Ab Ab
[23:55:27] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:Ab Ab Ab Ab Ab Ab Ab Ab Ab
[23:55:27] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:Ab Ab Ab Ab Ab Ab Ab Ab Ab
[23:55:27] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:Ab Ab Ab Ab Ab Ab Ab Ab Ab
[23:55:27] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:Ab Ab Ab Ab Ab Ab Ab Ab Ab
[23:55:27] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:Ab Ab Ab Ab Ab Ab Ab Ab Ab
[23:55:27] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:Ab Ab Ab Ab Ab Ab Ab Ab Ab
[23:55:27] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:Ab Ab Ab Ab Ab Ab Ab Ab Ab
[23:55:27] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:Ab Ab Ab Ab Ab Ab Ab Ab Ab
[23:55:27] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:Ab Ab Ab Ab Ab Ab Ab Ab Ab
[23:55:27] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:Ab Ab Ab Ab Ab Ab Ab Ab Ab
[23:55:27] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:Ab Ab Ab Ab Ab Ab Ab Ab Ab
[23:55:27] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:Ab Ab Ab Ab Ab Ab Ab Ab Ab
[23:55:27] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:Ab Ab Ab Ab Ab Ab Ab Ab Ab
[23:55:27] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:Ab Ab Ab Ab Ab Ab Ab Ab Ab
[23:55:27] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:Ab Ab Ab Ab Ab Ab Ab Ab Ab
[23:55:27] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:Ab Ab Ab Ab Ab Ab Ab Ab Ab
[23:55:27] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:Ab Ab Ab Ab Ab Ab Ab Ab Ab
[23:55:27] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:Ab Ab Ab Ab Ab Ab Ab Ab Ab
[23:55:27] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:Ab Ab Ab Ab Ab Ab Ab Ab Ab
[23:55:27] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:Ab Ab Ab Ab Ab Ab Ab Ab Ab
[23:55:27] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:Ab Ab Ab Ab Ab Ab Ab Ab Ab
[23:55:27] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:Ab Ab Ab Ab Ab Ab Ab Ab Ab
[23:55:27] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:Ab Ab Ab Ab Ab Ab Ab Ab Ab
[23:55:27] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:Ab Ab Ab Ab Ab Ab Ab Ab Ab
[23:55:27] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:Ab Ab Ab Ab Ab Ab Ab Ab Ab
[23:55:27] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:Ab Ab Ab Ab Ab Ab Ab Ab Ab
[23:55:27] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:Ab Ab Ab Ab Ab Ab Ab Ab Ab
[23:55:27] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:Ab Ab Ab Ab Ab Ab Ab Ab Ab
[23:55:27] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:Ab Ab Ab Ab Ab Ab Ab Ab Ab
[23:55:27] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:Ab Ab Ab Ab Ab Ab Ab Ab Ab
[23:55:27] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:Ab Ab Ab Ab Ab Ab Ab Ab Ab
[23:55:27] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:Ab Ab Ab Ab Ab Ab Ab Ab Ab
[23:55:27] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:Ab Ab Ab Ab Ab Ab Ab Ab Ab
[23:55:27] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:Ab Ab Ab Ab Ab Ab Ab Ab Ab
[23:55:27] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:Ab Ab Ab Ab Ab Ab Ab Ab Ab
[23:55:27] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:Ab Ab Ab Ab Ab Ab Ab Ab Ab
[23:55:27] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:Ab Ab Ab Ab Ab Ab Ab Ab Ab
[23:55:27] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:Ab Ab Ab Ab Ab Ab Ab Ab Ab
[23:55:27] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:Ab Ab Ab Ab Ab Ab Ab Ab Ab
[23:55:27] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:Ab Ab Ab Ab Ab Ab Ab Ab Ab
[23:55:27] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:Ab Ab Ab Ab Ab Ab Ab Ab Ab
[23:55:27] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:Ab Ab Ab Ab Ab Ab Ab Ab Ab
[23:55:27] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:Ab Ab Ab Ab Ab Ab Ab Ab Ab
[23:55:27] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:Ab Ab Ab Ab Ab Ab Ab Ab Ab
[23:55:27] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:Ab Ab Ab Ab Ab Ab Ab Ab Ab
[23:55:27] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:Ab Ab Ab Ab Ab Ab Ab Ab Ab
[23:55:27] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:Ab Ab Ab Ab Ab Ab Ab Ab Ab
[23:55:27] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:Ab Ab Ab Ab Ab Ab Ab Ab Ab
[23:55:27] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:Ab Ab Ab Ab Ab Ab Ab Ab Ab
[23:55:27] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:Ab Ab Ab Ab Ab Ab Ab Ab Ab
[23:55:27] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:Ab Ab Ab Ab Ab Ab Ab Ab Ab
[23:55:27] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:Ab Ab Ab Ab Ab Ab Ab Ab Ab
[23:55:27] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:Ab Ab Ab Ab Ab Ab Ab Ab Ab
[23:55:27] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:Ab Ab Ab Ab Ab Ab Ab Ab Ab
[23:55:27] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:Ab Ab Ab Ab Ab Ab 
```
As you can see, no spaces omitted.


Later, in the actual drawScreen() function in GUIManual, the following is run. 
```
bodyTexts = page.getBodyList();
```
Resulting in:
```
[23:51:30] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:33]: PEDEBUG:Ab Ab Ab Ab Ab Ab Ab Ab AbAb Ab Ab Ab Ab Ab Ab Ab AbAb Ab Ab Ab Ab Ab Ab Ab AbAb Ab Ab Ab Ab Ab Ab Ab AbAb Ab Ab Ab Ab Ab Ab Ab AbAb Ab Ab Ab Ab Ab Ab Ab AbAb Ab Ab Ab Ab Ab Ab Ab AbAb Ab Ab Ab Ab Ab Ab Ab AbAb Ab Ab Ab Ab Ab Ab Ab AbAb Ab Ab Ab Ab Ab Ab Ab AbAb Ab Ab Ab Ab Ab Ab Ab AbAb Ab Ab Ab Ab Ab Ab Ab AbAb Ab Ab Ab Ab Ab Ab Ab AbAb Ab Ab Ab Ab Ab Ab Ab AbAb Ab Ab Ab Ab Ab Ab Ab AbAb Ab Ab Ab Ab Ab Ab Ab Ab
[23:51:30] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:Ab Ab Ab Ab Ab Ab Ab Ab
[23:51:30] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:AbAb Ab Ab Ab Ab Ab Ab Ab
[23:51:30] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:AbAb Ab Ab Ab Ab Ab Ab Ab
[23:51:30] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:AbAb Ab Ab Ab Ab Ab Ab Ab
[23:51:30] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:AbAb Ab Ab Ab Ab Ab Ab Ab
[23:51:30] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:AbAb Ab Ab Ab Ab Ab Ab Ab
[23:51:30] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:AbAb Ab Ab Ab Ab Ab Ab Ab
[23:51:30] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:AbAb Ab Ab Ab Ab Ab Ab Ab
[23:51:30] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:AbAb Ab Ab Ab Ab Ab Ab Ab
[23:51:30] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:AbAb Ab Ab Ab Ab Ab Ab Ab
[23:51:30] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:AbAb Ab Ab Ab Ab Ab Ab Ab
[23:51:30] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:AbAb Ab Ab Ab Ab Ab Ab Ab
[23:51:30] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:AbAb Ab Ab Ab Ab Ab Ab Ab
[23:51:30] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:AbAb Ab Ab Ab Ab Ab Ab Ab
[23:51:30] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:AbAb Ab Ab Ab Ab Ab Ab Ab
[23:51:30] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:AbAb Ab Ab Ab Ab Ab Ab Ab
[23:51:30] [Client thread/INFO] [STDOUT]: [moze_intel.projecte.manual.AbstractPage:getBodyList:35]: PEDEBUG:Ab
```
Here spaces are omitted.

Conclusion: 
When run during postinit, getBodyList works fine.
When run in drawScreen(), the SECOND time it is run, it doesn't. First run works fine.

**Comment by Lilylicious** [comment]
It uses the function below to add newlines at the right spots, can someone help me decipher this to see if it could cause it to cause the error?

```    
String wrapFormattedStringToWidth(String p_78280_1_, int p_78280_2_)
    {
        int j = this.sizeStringToWidth(p_78280_1_, p_78280_2_);

        if (p_78280_1_.length() <= j)
        {
            return p_78280_1_;
        }
        else
        {
            String s1 = p_78280_1_.substring(0, j);
            char c0 = p_78280_1_.charAt(j);
            boolean flag = c0 == 32 || c0 == 10;
            String s2 = getFormatFromString(s1) + p_78280_1_.substring(j + (flag ? 1 : 0));
            return s1 + "\n" + this.wrapFormattedStringToWidth(s2, p_78280_2_);
        }
    }```

**Comment by Lilylicious** [comment]
https://github.com/Lilylicious/ProjectE/commit/2cdce9a8b785fe58e23ae8cc274bc9d0740e8db6 fixes it. Holy shit that was difficult to track down and fix. Over two hours since I started working on it proper, 162 additions and 3 deletions, all for changing one line in a private function.

The one thing that was causing the problem: 
```boolean flag = c0 == 32 || c0 == 10;```
needed to be
```boolean flag = c0 == 10;```

To be fair, that change introduced another "problem" that was easily fixed with
```
bodyTexts.get(i).charAt(0) == 32 ? bodyTexts.get(i).substring(1) : bodyTexts.get(i)
```
rather than
```
bodyTexts.get(i)
```

**Comment by Lilylicious** [comment]
Note to self: Look into embedded images.

**Comment by Lilylicious** [comment]
A bug I hadn't noticed before now: If you go to the next page if you're on the second to last page, and the last page is on the left, you get index out of bounds. I think the link leads to the problem area, but I'm not sure, suuuuper tired.

https://github.com/Lilylicious/ProjectE/blob/2cdce9a8b785fe58e23ae8cc274bc9d0740e8db6/src/main/java/moze_intel/projecte/gameObjs/gui/GUIManual.java#L107

**Comment by williewillus** [comment]
So I've done a bit of minor cleanup (on my fork): moved some things around to make more sense, removed some unused stuff, moved the init to the proxy so it doesn't crash servers, and constantified a lot of stuff in the GUI.
What I want to do next is to have a way to "bake" the index during init so it doesn't get bsaically recomputed every time the gui is reopened. Along the same lines, doing that would allow the index pages to become actual pages themselves, reducing the special support code needed for them in the gui. Another thing is linking subpages to their superpage somehow (through a field or something), instead of their relationship being defined in terms of their names only.
Also, I don't know if multi-language support has been tested, but along with the index "baking" above there needs to be a way to trigger a rebake/recompute of the index and subpages when the language is changed or the resource system is reloaded.

I'll be done (depending on my time) in 1-3 days. Yall could start taking a look at the lore itself (at the google doc link above).

**Comment by Lilylicious** [comment]
I haven't done any testing about the multi-language support, but I've given it some thoughts. We're inevitably going to run into a language where there's a word (or button text) that is longer than the GUI is wide (french). I was thinking maybe have something during init go through the localized strings and find the widest, then do scaling to make it fit.

That'd increase launch time though, so maybe do the scaling for each individual page when rendering it? That'd cause different pages to be scaled differently though. 

Regarding the subpages being unlinked to the superpages, that's something I intentionally left for you to clean up. Not too comfortable with the whole sub vs super relationship, figured I'd let you do it and see how you manage it. The current implementation works, but it's a weak link. Linking them without using the names would allow us to get rid of the _X's in the name of the subpages.

I'll be keeping an eye on what you get up to in your fork, partly for learning purposes and partly to see if I get any ideas. Other than that I'll also start looking at embedded images/hybrid image pages. Experimenting with multi-language support and scaling is another task I'll be looking into.

**Comment by williewillus** [comment]
@Lilylicious I'm done making all the tweaks I wanted. In short
* Index pages are now real pages, so no more negative pages. They're blank, and the buttons get rendered over them like before
* Concept of spreads
* Move everything clientside so it doesnt crash servers
* Just some minor cleanup here and there
* It automatically reorders and regenerates itself when the resource pack is switched or languages are changed
* Things actually seem to be working fine with foreign languages already. Tested French and Chinese and everything  reordered and looks fine.

**Comment by Lilylicious** [comment]
I'll be giving it a once-over again to see if anything pops out to me as something I can work on, otherwise it's about ready for pure content-focused work. Great job on it, this is really coming together nicely.

**Comment by Lilylicious** [comment]
@sinkillerj I'm ready to call this code complete, v1.0 of the manual. It is possible that at some point in the future I'll fancify it a bit and/or add/remove things, but the first version is code complete. 

In other words, code is complete, ready for content. When content is in, I think it's ready to be deployed.

**Comment by Lilylicious** [comment]
https://github.com/Lilylicious/ProjectE/tree/manual is the branch that's ready. Could make a pull for it if you'd like, could also wait with that until after content.

**Comment by sinkillerj** [comment]
Eh go ahead and make the pull, will be easier to review
